library/net.cpp

```diff
@@ -1410,3 +1412,96 @@ int dnet_trans_forward(struct dnet_io_req *r, struct dnet_net_state *orig, struc
 
 	return dnet_trans_send(t, r);
 }
+
+dnet_io_req *n2_create_io_req(struct n2_call *n2_c) {
+	// To be freed in C code: used C-style allocation
+	auto r = static_cast<dnet_io_req *>(calloc(1, sizeof(dnet_io_req)));
+	if (!r) {
+		throw std::logic_error("Sending to local state: no memory");
+	}
+
+	r->n2_c = n2_c;
+	return r;
+}
+
+void n2_schedule_io_req(struct dnet_net_state *st, struct dnet_io_req *r, struct dnet_access_context *context) {
+	r->context = dnet_access_context_get(context);
+
+	clock_gettime(CLOCK_MONOTONIC_RAW, &st->rcv_start_ts);
+	st->rcv_finish_ts = st->rcv_start_ts;
+
+	r->st = dnet_state_get(st);
+	dnet_schedule_io(st->n, r);
+}
+
+void n2_send_response_local(struct dnet_net_state *st, struct n2_call *n2_c,
+                            struct n2_message *n2_msg, struct dnet_access_context *context) {
+	DNET_LOG_DEBUG(st->n, "Sending response to local state"); // TODO: dump response info
+
+	auto r = n2_create_io_req(n2_c);
+	n2_msg->make_owning();
+	n2_c->reply(n2_msg);
+	n2_schedule_io_req(st, r, context);
+}
+
+void n2_send_error_local(struct dnet_net_state *st, struct n2_call *n2_c,
+                         int errc, struct dnet_access_context *context) {
+	DNET_LOG_DEBUG(st->n, "Sending error response to local state: %d", errc);
+
+	auto r = n2_create_io_req(n2_c);
+	n2_c->reply_error(errc);
+	n2_schedule_io_req(st, r, context);
+}
+
+int n2_send_response(struct dnet_net_state *st,
+                     struct n2_call *n2_c,
+                     struct n2_message *n2_msg,
+                     struct dnet_access_context *context) {
+	try {
+		dnet_cmd *cmd = &n2_msg->cmd;
+		DNET_LOG(st->n, DNET_LOG_NOTICE, "%s: %s: reply trans: %lld -> %s (%p): cflags: %s",
+			 dnet_dump_id(&cmd->id), dnet_cmd_string(cmd->cmd), (unsigned long long)cmd->trans,
+			 dnet_state_dump_addr(st), static_cast<void *>(st),
+			 dnet_flags_dump_cflags(cmd->flags));
+
+		if (st == st->n->st) {
+			/*
+			 * If destination is local node, call with reply are scheduled to the proper IO pool directly.
+			 * If n2_msg hold fd instead of data it should be read here (see n2_io_req_local).
+			 */
+			n2_send_response_local(st, n2_c, n2_msg, context);
+
```

Comment:
It doesn't so matter, but `} else {` and `} catch (...) {` already separate code blocks. These empty lines look unnecessary.